### PR TITLE
Methods to fix perms of managed dirs

### PR DESCRIPTION
**Pull request by ericfranz**
Add two methods to manage permissions of managed directories: fix_dataset_root_permissions and fix_app_permissions

These two methods would be called from a Rake task or from a user initiated button click in the web app. The result of the methods is a log hash that indicates which ones were updated and 

This code would probably be made simpler if there existed Dataset and App models (or even ManagedFile and ManagedDir models). Then each would know how to "fix itself" and whether they were recently updated or an error occurred (ActiveRecord/ActiveModel has mechanisms already to store all this information).

Also not sure how to write tests for this.

### CODE REVIEW

File: app/models/managed_file.rb [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  #
+  # OodSupport::Nfs4Entry#group_owner_entry? is broken cause it assumes
+  # principle is GROUP AND it will contain the g
+  #
+  def facls_different?(acl1, acl2)
+    sanitize_acl_for_comparison(acl1) != sanitize_acl_for_comparison(acl2)
+  end
+
+  # remove g from A:g:GROUP line
+  # remove o from A::OWNER@: line
```
Comment (MorganRodgers) [inline app/models/managed_file.rb]: `# remove o from A::OWNER@: line`

Out of date comment?

**Review by MorganRodgers — COMMENTED** [review]
Over all I think that this works.

Since we are managing these files specifically to ensure that they can be shared, and you have chosen to hardcode the desired permissions, consider (but you don't have to of course) renaming ManagedFile to ManagedSharedFile. One thing that I would like to see done slightly differently is for ManagedFile to make no assumptions about which files it should manage so something like:

```
def fix_app_permissions(apps)
  # do stuff with apps
end

def fix_dataset_permissions(datasets)
  # do stuff with datasets
end
```

As for tests I would confirm that:

- `Managed(Shared)?File#facls_different?` works as expected with exactly the same, rule-equivalent, and rule-different ACLs
- `ManagedFile#fix_(app|dataset)_permissions` returns the correct hash if a mock'd Open3.capture3 succeeds or fails with a given list of apps

**Comment by ericfranz** [comment]
Changed fix_dataset_root_permissions and fix_app_permissions to accept arguments. It still uses `Mapping.users_that_have_mappings_to_dataset` and `Mapping.users_that_have_mappings_to_app` to get the list of users that should have access to a given path.

If I were to improve this to either make it more generic (useful in other contexts, like with TotalSim) or make it more function to require no mocking while writing tests, there are two paths:

1. Change the fix methods to accept block that accepts a single argument "path" and returns an array of users that should have access to that path. Still even with this the method signatures are a bit odd. One asks for all of the paths we will be applying a particular facl template to, the other asks for a root directory that we will recursively update all the facls of files and directories in. Not sure how that translates to more generic.
2. Switch to an object oriented approach using what are a type of command object: SharedFile, SharedDirectory, SharedApp, SharedDataset are example classes which contain a `path`, a single method `fix_permissions`. Each object has its own facl template (and thus knows how to fix its own permissions). SharedApp and SharedDataset could be just instances of UserRestrictedDirectory where permissions are 750 and when initializing this instance the list of users is provided to the constructor. This shifts any complexity in just initializing the list of the directory objects (which are either SharedApp/SharedDataset/UserRestrictedDirectory, or SharedFile, or SharedDirectory) and then over that list doing something like: `dirs.map(&:fix_permissions)` which gives you that list of files whose permissions were actually modified, and any errors. Creating the list of directory objects to manage ends up being the domain specific code (specific to shiny admin or to total sim etc.).

That said I think its to a place where I can write two tests for it and verify so we can revisit this later.

**Review by MorganRodgers — APPROVED** [review]
I like the return of `FaclChangeReport` from the `fix_*_permission` methods.

**Pull request merged**